packages/labs/src/components/timezone-select/timezoneSelect.tsx

```diff
@@ -0,0 +1,509 @@
+                resetOnClose={true}
+                popoverProps={finalPopoverProps}
+                disabled={disabled}
+                onQueryChange={this.handleQueryChange}
+            >
+                {this.renderTarget()}
+            </TypedSelect>
+        );
+    }
+
+    public componentWillReceiveProps(nextProps: ITimezoneSelectProps) {
+        const nextDate = nextProps.date || new Date();
+        const dateChanged = this.state.date.getTime() !== nextDate.getTime();
+
+        if (dateChanged) {
+            this.updateTimezones(nextDate);
+        }
+        if (dateChanged || this.props.showLocalTimezone !== nextProps.showLocalTimezone) {
+            this.initialTimezones = getInitialTimezoneItems(nextDate, nextProps.showLocalTimezone);
+        }
+
+        const nextState: ITimezoneSelectState = {};
+        if (dateChanged) {
+            nextState.date = nextDate;
+        }
+        if (this.state.value !== nextProps.value) {
+            nextState.value = nextProps.value;
+        }
+        this.setState(nextState);
+    }
+
+    private renderTarget() {
+        const {
+            disabled,
+            targetRenderer,
+            targetDisplayFormat = TimezoneDisplayFormat.OFFSET,
+            defaultValue,
+            defaultToLocalTimezone,
+            placeholder,
+        } = this.props;
+        const { date, value } = this.state;
+
+        const finalDefaultValue = defaultValue || (defaultToLocalTimezone && getLocalTimezone());
+
+        const finalPlaceholder = placeholder !== undefined
+            ? placeholder
+            : formatTimezone(PLACEHOLDER_TIMEZONE, date, targetDisplayFormat);
+
+        const finalTargetRenderer = targetRenderer || this.defaultTargetRenderer;
+        return finalTargetRenderer({
+            date,
+            defaultDisplayValue: formatTimezone(finalDefaultValue, date, targetDisplayFormat),
+            defaultValue: finalDefaultValue,
+            disabled,
+            displayValue: formatTimezone(value, date, targetDisplayFormat),
+            placeholder: finalPlaceholder,
+            value,
+        });
+    }
+
+    private defaultTargetRenderer: ITimezoneSelectTargetRenderer = (targetProps) => {
+        const { targetClassName } = this.props;
+        const { value, displayValue, defaultDisplayValue, placeholder, disabled } = targetProps;
+        const isPlaceholder = !value;
+        const classes = classNames(Classes.TIMEZONE_SELECT_TARGET, targetClassName, {
+            [Classes.TIMEZONE_SELECT_TARGET_PLACEHOLDER]: isPlaceholder,
+        });
+
+        return (
+            <Button
+                className={classes}
+                text={displayValue || defaultDisplayValue || placeholder}
+                rightIconName="caret-down"
+                disabled={disabled}
+            />
+        );
+    }
+
+    private updateTimezones(date: Date): void {
+        const timezones = getTimezoneItems(date);
+        this.timezones = timezones;
```

Comment:
```this.timezones = getTimezoneItems(date);```

why two lines?

Reply:
`timezones` is used on the next line
so, the ordering dependency felt clearer this way
otherwise, if you swap the two lines, its a bug

Reply:
fair enough